before_install_linux.sh

```diff
@@ -15,35 +15,50 @@ if [ ! -d /usr/include/freetype ]; then
   sudo ln -sf /usr/include/freetype2 /usr/include/freetype
 fi
 
+dir_options=""
+if [ -v CONFIGURE_OPTIONS ]; then
+  dir_options=$CONFIGURE_OPTIONS
+fi
+
 if [ -v IMAGEMAGICK_VERSION ]; then
-  version=(${IMAGEMAGICK_VERSION//./ })
-  if (( "${version[0]}${version[1]}" >= 69 )); then
-    wget https://github.com/ImageMagick/ImageMagick6/archive/${IMAGEMAGICK_VERSION}.tar.gz
-    tar -xzf ${IMAGEMAGICK_VERSION}.tar.gz
-    cd ImageMagick6-${IMAGEMAGICK_VERSION}
+  if [ -d build-ImageMagick/ImageMagick-${IMAGEMAGICK_VERSION}-${dir_options} ]; then
+    cd build-ImageMagick/ImageMagick-${IMAGEMAGICK_VERSION}-${dir_options}
   else
-    wget http://www.imagemagick.org/download/releases/ImageMagick-${IMAGEMAGICK_VERSION}.tar.xz
-    tar -xf ImageMagick-${IMAGEMAGICK_VERSION}.tar.xz
-    cd ImageMagick-${IMAGEMAGICK_VERSION}
+    mkdir -p build-ImageMagick
+
+    version=(${IMAGEMAGICK_VERSION//./ })
+    if (( "${version[0]}${version[1]}" >= 69 )); then
+      wget https://github.com/ImageMagick/ImageMagick6/archive/${IMAGEMAGICK_VERSION}.tar.gz
+      tar -xf ${IMAGEMAGICK_VERSION}.tar.gz
+      mv ImageMagick6-${IMAGEMAGICK_VERSION} build-ImageMagick/ImageMagick-${IMAGEMAGICK_VERSION}-${dir_options}
+    else
+      wget http://www.imagemagick.org/download/releases/ImageMagick-${IMAGEMAGICK_VERSION}.tar.xz
+      tar -xf ImageMagick-${IMAGEMAGICK_VERSION}.tar.xz
+      mv ImageMagick-${IMAGEMAGICK_VERSION} build-ImageMagick/ImageMagick-${IMAGEMAGICK_VERSION}-${dir_options}
+    fi
+
+    cd build-ImageMagick/ImageMagick-${IMAGEMAGICK_VERSION}-${dir_options}
+
+    options="--with-magick-plus-plus=no --disable-docs"
+    if [ -v CONFIGURE_OPTIONS ]; then
+      options="${CONFIGURE_OPTIONS} $options"
+    fi
+
+    CC="ccache cc" CXX="ccache c++" ./configure --prefix=/usr $options
+
+    make -j
   fi
+
+  sudo make install -j
+  cd ../..
 else
   echo "you must specify an ImageMagick version."
```

Comment:
Maybe we should move the whole creation of the folder and building ImageMagick inside a function instead? We can then do early exists instead of an `else` very far away from the `if` statement.

Reply:
Seems it is needed to work correctly for Travis.

If it does not back, CI script does not work as expected.
(I don’t know why needed)

Reply:
I think it might work as a separate function if we do something like this:

```bash
build_imagemagick () {
  mkdir -p build-ImageMagick

  ...
}

if [ ! -d $build_dir ]; then
  build_imagemagick
fi

cd $build_dir
sudo make install -j
cd $project_dir
```

We'll need absolute `$project_dir` and `$build_dir` variables to make sure we end up in the right location.